clarity/src/vm/database/tests/clarity_db_tests.rs

```diff
@@ -0,0 +1,177 @@
+use fake::{Faker, Fake};
+use rusqlite::NO_PARAMS;
+use stacks_common::util::hash::Sha512Trunc256Sum;
+
+use crate::vm::{contracts::Contract, database::{clarity_store::ContractCommitment, ClarityBackingStore, ClarityDatabase, ClaritySerializable, MemoryBackingStore, NULL_BURN_STATE_DB, NULL_HEADER_DB}, fakes::raw::EnglishWord, Value};
+
+#[test]
+fn insert_contract() {
+    for _ in 0..1000 {
```

Comment:
Would it be possible to use `fake` through `proptest`? `proptest` automatically re-runs the test for you, so the `loop` can be removed. Depending on how you will define the strategy, you may have the option to shrink any errors found. Replaying a failed test is another important quality of `proptest`.

Reply:
I'm not so familiar with `proptest` so if there's a good way of doing this I'm all ears! The difference being that when you call `Faker.fake()` you instead call `Faker.fake_with_rng(&rng)` where you provide the `Rng` instance to use.

But yeah, the loop is there just to ensure that instances of varying levels of complexity are tested otherwise.

Reply:
@moodmosaic does `proptest` have any way to provide a generator-function? i.e. could I have a function `fn some_test(contract: Contract)` and specify how `Contract` should be created (and ideally with an `Rng` instance)?

Reply:
>a generator-function

Proptest uses a bit different terminology than most other Haskell QuickCheck-clones (although, it's debatable whether Proptest is actually a Haskell QuickCheck-clone as it's heavily inspired by Python's Hypothesis and *Elrnag*'s QuickCheck).

The way most, if not all, property-based testing libraries work is be allowing you to create a pair of a generator and a shrinker. A generator knows how to produces values of type `a` and a shrinker knows how to reduce `a` into an list of of smaller `a`s (to simplify those values and attempt to identify minimal failing cases).

While the original Haskell QuickCheck (and clones) name such a pair Arbitrary, Proptest uses the name Strategy.

Given the context above, to answer the questions:

>could I have a function `fn some_test(contract: Contract)` and specify how `Contract` should be generated

Proptest allows for the creation of custom Strategies to generate (and shrink) any type, including `Contract`. Although the test runner may not automatically infer how to instantiate a function like `fn some_test(contract: Contract)`, you can define a Strategy for `Contract` and utilize it within a standard test using proptest.

>and ideally pass an existing `Rng` instance

You could theoretically [get the prng from proptest](https://altsysrq.github.io/rustdoc/proptest/0.9.1/proptest/test_runner/struct.TestRunner.html#method.rng) and [pass it in fake](https://docs.rs/fake/latest/fake/trait.Fake.html#method.fake_with_rng), however this approach has implications:
1. You're required to always implement a shrinker since Proptest won't inherently understand how to shrink data generated by fake.
2. This might affect the randomness distribution of your tests. Proptest aims to generate test cases that are uniformly distributed across the space of possible inputs defined by the strategies, which is crucial for the effectiveness of property-based testing. It's unclear if integrating fake in this manner maintains this distribution, a key factor that can be assessed using benchmarks like the [dieharder](https://linux.die.net/man/1/dieharder) tests.

At the end, the effort must be done as in [`fake::locales::Data`](https://docs.rs/fake/latest/fake/locales/trait.Data.html), and as @jcnelson mentions in https://github.com/stacks-network/stacks-core/pull/4477#issuecomment-1977076270, and even though feasible, might be best concentrated within a single tool to maintain consistency and the integrity of the testing process.

Just using Proptest alone for both generation and shrinking is advisable for simplicity and reliability...

Reply:
I'm exploring with Anthony's help moving all generation to `proptest` - it hasn't been trivial, but there was a lot of work already done in the `clarity-wasm` repo for Clarity types. Will get back on this PR once I've got a better grasp of the effort needed and how well it works for "regular" unit tests.